add tests to increase enum serialization code coverage

Jacksondr5 opened the pull request:
Opening this PR because I messed up #39561, addressed feedback by @Jozkee 
Helps address #32341
Adds coverage to branch where WriteWithQuotes goes over the name cache soft limit. code
Adds coverage to branches where a Dictionary key is a numeric enum that is not in the cache and the value is not a valid identifier. code

Jacksondr5 committed:
add tests to increase enum serialization code coverage

Jozkee commented on src/libraries/System.Text.Json/tests/Serialization/EnumConverterTests.cs:
```diff
@@ -294,5 +295,222 @@ public enum MyEnum
+        {
+            // Ensure we don't throw OutOfMemoryException.
+
+            var options = new JsonSerializerOptions
+            {
+                Converters = { new JsonStringEnumConverter() }
+            };
```
You don't have to use `JsonSerializerOptions` in this case. 

For `VeryLargeAmountOfEnumsToSerialize` the options are used because we don't serialize as string the enum values by default, you need to opt-in for it passing the `JsonStringEnumConverter`.

We do serialize as string by default the enum values when they are dictionary keys.

Jacksondr5 replied:
Fixed 👍

jozkee reviewed (approved):
Thanks for the PR.
Could you please add cover for these few remaining lines as well? You can take a look at `MoreThan64EnumValuesToSerialize` test to have an idea on how to do so.
https://github.com/dotnet/runtime/blob/e0df256a4458364937e524aeee3f525c71330e0c/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs#L357-L360

Otherwise, lgtm.

Jacksondr5 committed:
implement feedback and add test to cover naming policy scenario

Jacksondr5 commented:
Thanks for the feedback!  I looked at the lines you mentioned (357-360) and they appear to be covered already:
![image](https://user-images.githubusercontent.com/9014034/92998782-0c0e2680-f4ea-11ea-9dab-7562d0de9e8c.png)
Is there something I'm missing here?

I did add a test for a branch that wasn't covered where the cache soft limit is hit with a naming policy provided.  It's the false branch here:
https://github.com/dotnet/runtime/blob/4bc91cb3805bc4a119fc18f4a32d5126ac72db60/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs#L193-L196

Jozkee commented:
@Jacksondr5 that's odd, copying your changes locally, the lines aforementioned appear as uncovered on my end. 

![image](https://user-images.githubusercontent.com/16040868/93132820-229fb380-f68b-11ea-9f43-c9c1b8fba9b8.png)

Jozkee commented on src/libraries/System.Text.Json/tests/Serialization/EnumConverterTests.cs:
```diff
@@ -232,6 +233,23 @@ public static void MoreThan64EnumValuesToSerialize()
             }
         }
 
+        [Fact]
+        public static void MoreThan64EnumValuesToSerializeWithNamingPolicy()
+        {
+            var options = new JsonSerializerOptions
+            {
+                Converters = { new JsonStringEnumConverter(new ToLower()) }
+            };
+
+            for (int i = 0; i < 128; i++)
+            {
+                MyEnum value = (MyEnum)i;
+                string asStr = value.ToString().ToLowerInvariant();
+                string expected = char.IsLetter(asStr[0]) ? $@"""{asStr}""" : asStr;
+                Assert.Equal(expected, JsonSerializer.Serialize(value, options));
+            }
+        }
```
Does this test covers `EnumConverter<T>.WriteWithQuotes`? I was expecting to see a `Dictionary<MyEnum, int>` since I think that's the scenario that uses the `WriteWithQuotes` method.

Here's the test that I had in mind:
```cs
[Fact]
public static void MoreThan64EnumDictionaryKeysToSerialize()
{
    for (int i = 0; i < 128; i++)
    {
        MyEnum myEnum = (MyEnum)i;
        var dictionary = new Dictionary<MyEnum, int> { { myEnum, i } };
        string expected = $@"{{""{myEnum}"":{i}}}";
        Assert.Equal(expected, JsonSerializer.Serialize(dictionary));
    }
}
```

Jacksondr5 replied:
This test actually covers `EnumConverter<T>.Write`.  The false condition of this code wasn't being visited by any tests (although given that we're running different test commands, maybe there's something I'm doing wrong here 😉 )
https://github.com/dotnet/runtime/blob/4bc91cb3805bc4a119fc18f4a32d5126ac72db60/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs#L193-L196

Sorry for the confusion.  While I was looking at the WriteWithQuotes branch you were talking about, I came across this other piece of uncovered code.

ahsonkhan commented:
> Is there something I'm missing here?

> that's odd, copying your changes locally, the lines aforementioned appear as uncovered on my end.

Howa are you running the tests and measuring coverage (what command)? Are both of you running the tests in the same way (same architecture/runtime version/OS)? Maybe one of you has outerloop tests on.

Jacksondr5 commented:
I was running these using `dotnet build /t:Test /p:Coverage=true /p:Outerloop=true` and the branch is visited 2097088 times.  Comparing the numbers on my output to the screenshot @Jozkee provided, it does seem like outerloop could be the difference.

Jozkee commented:
@Jacksondr5 @ahsonkhan Ah, that's right, I was missing the OuterLoop argument. Since the lines that I pointed out are already covered by OuterLoop tests; this looks good to me as is.

These few lines are the only lines that are missing coverage on `EnumConverter.cs`:

![image](https://user-images.githubusercontent.com/16040868/93386830-23684f00-f81d-11ea-9648-be340a14b176.png)

For these yellow lines, I think we should instead address the comment given that https://github.com/dotnet/runtime/issues/29000 is already closed.

![image](https://user-images.githubusercontent.com/16040868/93386977-5ad6fb80-f81d-11ea-9818-0175e9a3a860.png)

Jacksondr5 committed:
remove unnecessary bound check now that all primitive types are supported

Jacksondr5 commented:
Fixed the yellow lines.  Since the coverage tool is currently acting weird with switch expressions, I blew the one below out into a bunch of if statements so that I could verify that all branches were being hit.  I didn't commit those changes though, they were just for testing.
![image](https://user-images.githubusercontent.com/9014034/93411245-b4184c80-f868-11ea-8fa2-c265d2e31a7e.png)


As for the unreached code in the constructor, I've got a question about that.  To test that the cache limit works (ie doesn't throw an OutOfMemoryException), the other tests iterate over a flag enum and call Serialize on each iteration.  I don't think we can do that with the constructor since it only caches the specifically defined values.  Given that the other tests that cover the cache limit serialize up to 2097152 enums, I think we would need to have a hilariously large enum literal to truly test that the constructor code doesn't throw an OutOfMemoryException.  Maybe I could do something with reflection, but I wanted to check before I wrote anything.  Is there some other way to approach testing this code?

Jozkee commented:
> As for the unreached code in the constructor, I've got a question about that. To test that the cache limit works (ie doesn't throw an OutOfMemoryException), the other tests iterate over a flag enum and call Serialize on each iteration

@Jacksondr5 you can cover those lines by just serializing an `enum` with 65 names, that will make the constructor to stop caching names at the soft limit (64). Now that I say that out loud, I think that maybe writing such test is not even worth it, we don't have to cover every single line just to strive to have as much lines covered as possible; 100% coverage in this file is already impossible due to unreachable code.

@layomia is there any value on covering the soft limit check in the constructor? if so, could you suggest a test for it?

> Since the coverage tool is currently acting weird with switch expression, I blew the one below out into a bunch of if statements so that I could verify that all branches were being hit.

I'm glad that you confirmed that those lines are actually being covered but IMO I don't think we should make product changes (change source files) in PRs like this where the goal is add test coverage.

> Fixed the yellow lines. 

As I said, I think this PR should not contain product changes and we should open a separate PR that addresses the TODO comment.

Jacksondr5 committed:
Revert "remove unnecessary bound check now that all primitive types are supported"

This reverts commit 0a6a345ed3eaf65be8efbb51c6258a8448aaf9c1.

Jacksondr5 commented:
Reverted all changed to EnumConverter.cs.  I can open up a PR to address the TODO comment if desired.

Jozkee commented on src/libraries/System.Text.Json/tests/Serialization/EnumConverterTests.cs:
```diff
@@ -232,6 +233,23 @@ public static void MoreThan64EnumValuesToSerialize()
+        {
+            var options = new JsonSerializerOptions
+            {
+                Converters = { new JsonStringEnumConverter(new ToLower()) }
```
Can you please rename class `ToLower` to `ToLowerNamingPolicy`.

Jozkee commented on src/libraries/System.Text.Json/tests/Serialization/EnumConverterTests.cs:
```diff
@@ -294,5 +312,217 @@ public enum MyEnum
+            [Fact]
+            public void SerilizeDictionaryWhenCacheIsFull()
+            {
+                var options = new JsonSerializerOptions
+                {
+                    Converters = { new JsonStringEnumConverter() }
+                };
```
I think this is not needed.

jozkee reviewed (approved):
Otherwise; lgtm.

Jozkee commented:
> I can open up a PR to address the TODO comment if desired.

@Jacksondr5 feel free to do so, thanks!

Jacksondr5 committed:
implement review feedback

remove unnecessary serializer options in test
rename JsonNamingPolicy ToLower -> ToLowerNamingPolicy

jozkee committed:
Remove commented-out code

ahsonkhan commented on src/libraries/System.Text.Json/tests/Serialization/EnumConverterTests.cs:
```diff
@@ -294,5 +312,212 @@ public enum MyEnum
             V = 1 << 21,
         }
+
+        [Fact, OuterLoop]
+        public static void VeryLargeAmountOfEnumDictionaryKeysToSerialize()
```
How long does this test end up taking (and roughly how much memory does it use up)?

Jozkee replied:
@Jacksondr5 Correct me if I'm wrong but I think it instantiates 2,097,152 (`1 << 21`) dictionaries. 

Is there something that we could do to reduce this number while keeping the coverage? Perhaps you could do `Serialize<MyEnum>` until you have the cache at the desired number and then call `Serialize<Dictionary<MyEnum, int>` on the desired edge cases.

Jacksondr5 replied:
I'll try my hand at benchmark.net and post some comparisons between `VeryLargeAmountOfEnumsToSerialize` and `VeryLargeAmountOfEnumDictionaryKeysToSerialize` so we can understand how long this takes and how serializing the `Dictionary<MyEnum,int>` compares to serializing `MyEnum`s.

ahsonkhan replied:
A rough estimate is sufficient (which probably doesn't need BDN, but feel free to go down that route). 

If you run the tests locally from master, and then compare the test time of your branch (remove the outerloop from the attribute to try it), that should give you a general idea. I am looking for an order of magnitude. If the tests went from taking ~1 minute, to 2+, or if we are using up 100s of MB/GB of memory (on Windows, Task manager could show it), then the test is too resource intensive, and we would definitely keep it as outerloop, and try to make it simpler, if possible.

If not, then outerloop is good enough. I asked to understand whether we need this test to be outerloop or if it is fast enough to become part of innerloop.

Jozkee replied:
Base time running `dotnet build /t:test`: 
```
=== TEST EXECUTION SUMMARY ===
    System.Text.Json.Tests  Total: 8715, Errors: 0, Failed: 0, Skipped: 0, Time: 33.520s
```

Time running after adding this new test:
```
=== TEST EXECUTION SUMMARY ===
    System.Text.Json.Tests  Total: 8716, Errors: 0, Failed: 0, Skipped: 0, Time: 40.564s
```

@ahsonkhan I would consider keeping this test as `OuterLoop` considering that the test above, `VeryLargeAmountOfEnumsToSerialize`, which uses a similar code minus the insane amount of dictionary instances, is 
 already signaled to run on `OuterLoop`.

ahsonkhan replied:
> Time: 33.520s
> Time: 40.564s

> I would consider keeping this test as `OuterLoop` considering that the test above, `VeryLargeAmountOfEnumsToSerialize`, which uses a similar code minus the insane amount of dictionary instances, is already signaled to run on `OuterLoop`.

Agreed. That sounds good.
